File: cpp/bubbleSort.cpp
#include <bits/stdc++.h>
#include <chrono>
using namespace std;
using namespace chrono;

// A function to implement insertion sort
void insertionSort(vector<int>& arr) {
    int n = arr.size();
    for (int i = 1; i < n; i++) {
        int key = arr[i];
        int j = i - 1;

        // Move elements of arr[0..i-1], that are greater than key,
        // to one position ahead of their current position
        while (j >= 0 && arr[j] > key) {
            arr[j + 1] = arr[j];
            j = j - 1;
        }
        arr[j + 1] = key;
    }
}

// Function to print an array to a file
void printArrayToFile(const vector<int>& arr, const string& filename) {
    ofstream outFile(filename);
    for (int num : arr) {
        outFile << num << endl;
    }
    outFile.close();
}

// Function to read an array from a file
vector<int> readArrayFromFile(const string& filename) {
    ifstream inFile(filename);
    vector<int> arr;
    int temp;
    while (inFile >> temp) {
        arr.push_back(temp);
    }
    inFile.close();
    return arr;
}

// Driver code
int main() {
    string inputFilename = "unsorted_numbers.txt";
    string outputFilename = "sorted_numbers.txt";

    auto start = high_resolution_clock::now();

    vector<int> arr = readArrayFromFile(inputFilename);
    insertionSort(arr);
    printArrayToFile(arr, outputFilename);

    auto stop = high_resolution_clock::now();
    auto duration = duration_cast<milliseconds>(stop - start);

    cout << "Total time taken: " << duration.count() << " milliseconds\n";
    return 0;
}

File: python/bubbleSort.py
import time

def bubble_sort(arr):
    n = len(arr)
    for i in range(n - 1):
        for j in range(n - i - 1):
            if arr[j] > arr[j + 1]:
                arr[j], arr[j + 1] = arr[j + 1], arr[j]

def read_array_from_file(filename):
    with open(filename, 'r') as file:
        arr = [int(line.strip()) for line in file.readlines()]
    return arr

def write_array_to_file(arr, filename):
    with open(filename, 'w') as file:
        for number in arr:
            file.write(f"{number}\n")

# Driver code
input_filename = 'unsorted_numbers.txt'
output_filename = 'sorted_numbers.txt'

start_time = time.time()

arr = read_array_from_file(input_filename)
bubble_sort(arr)
write_array_to_file(arr, output_filename)

end_time = time.time()
total_time = end_time - start_time

print(f"Total time taken: {total_time:.6f} seconds")
